File: backend/routers/vacaciones.py
"""Router de vacaciones. empresa_id: lecturas por X-Empresa-Id; escrituras heredadas del empleado en el service."""
from typing import Literal, Optional
from uuid import UUID

from fastapi import APIRouter, Depends, Query, Request
from fastapi.responses import Response

from schemas.vacaciones import (
    SaldoVacacionesResponse,
    SolicitudVacacionesCreate,
    SolicitudVacacionesListResponse,
    SolicitudVacacionesResponse,
)
from services.vacaciones_service import VacacionesService
from utils.empresa import get_empresa_id
from utils.permisos import Accion, Seccion, require_permission

router = APIRouter()

SECCION = Seccion.VACACIONES


def _svc() -> VacacionesService:
    return VacacionesService()


@router.get("", response_model=SolicitudVacacionesListResponse, dependencies=[Depends(require_permission(SECCION, Accion.READ))])
async def list_vacaciones(
    request: Request,
    area_id: Optional[UUID] = Query(None),
    empleado_id: Optional[UUID] = Query(None),
    estado: Optional[str] = Query(None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
    service: VacacionesService = Depends(_svc),
) -> SolicitudVacacionesListResponse:
    u = request.state.user
    return service.get_all(u.get("id"), u.get("rol"), get_empresa_id(request), area_id, empleado_id, estado, page, page_size)


@router.get("/exportar", dependencies=[Depends(require_permission(SECCION, Accion.READ))])
async def exportar_vacaciones(request: Request, formato: Literal["pdf", "excel", "csv", "word"] = Query("excel"), area_id: Optional[UUID] = Query(None), empleado_id: Optional[UUID] = Query(None), estado: Optional[str] = Query(None), service: VacacionesService = Depends(_svc)) -> Response:
    u = request.state.user
    d = service.exportar(u.get("id"), u.get("rol"), get_empresa_id(request), formato, area_id, empleado_id, estado)
    return Response(content=d.content, media_type=d.media_type, headers={"Content-Disposition": f'attachment; filename="{d.filename}"'})


# /saldo/{id} debe ir ANTES de /{id} para evitar colisión de rutas
@router.get("/saldo/{empleado_id}", response_model=SaldoVacacionesResponse, dependencies=[Depends(require_permission(SECCION, Accion.READ))])
async def get_saldo(empleado_id: UUID, service: VacacionesService = Depends(_svc)) -> SaldoVacacionesResponse:
    return service.get_saldo(empleado_id)


@router.get("/empleado/{empleado_id}", response_model=SolicitudVacacionesListResponse, dependencies=[Depends(require_permission(SECCION, Accion.READ))])
async def list_vacaciones_empleado(empleado_id: UUID, service: VacacionesService = Depends(_svc)) -> SolicitudVacacionesListResponse:
    return service.get_by_empleado(empleado_id)


@router.get("/{id}", response_model=SolicitudVacacionesResponse, dependencies=[Depends(require_permission(SECCION, Accion.READ))])
async def get_vacacion(id: UUID, request: Request, service: VacacionesService = Depends(_svc)) -> SolicitudVacacionesResponse:
    return service.get_by_id(id, get_empresa_id(request))


@router.post("", response_model=SolicitudVacacionesResponse, status_code=201, dependencies=[Depends(require_permission(SECCION, Accion.WRITE))])
async def create_vacacion(
    request: Request,
    body: SolicitudVacacionesCreate,
    service: VacacionesService = Depends(_svc),
) -> SolicitudVacacionesResponse:
    return service.create(body, request.state.user.get("id", "system"), request.state.user.get("rol"))


@router.put("/{id}/cancelar", response_model=SolicitudVacacionesResponse, dependencies=[Depends(require_permission(SECCION, Accion.WRITE))])
async def cancel_vacacion(id: UUID, request: Request, service: VacacionesService = Depends(_svc)) -> SolicitudVacacionesResponse:
    return service.cancel(id, get_empresa_id(request), request.state.user.get("id", "system"), request.state.user.get("rol"))

File: frontend/services/vacaciones.ts
import type {
  SaldoVacaciones,
  SolicitudVacaciones,
  SolicitudVacacionesCreate,
  SolicitudVacacionesListResponse,
} from "@/types/vacaciones"
import { apiFetch, descargarArchivo, type FormatoExport } from "@/services/api"

export async function fetchVacaciones(
  empresaIdOverride?: string,
  areaId?: string,
  empleadoId?: string,
  estado?: string,
  page = 1,
  pageSize = 20,
): Promise<SolicitudVacacionesListResponse> {
  const params = new URLSearchParams({ page: String(page), page_size: String(pageSize) })
  if (areaId) params.set("area_id", areaId)
  if (empleadoId) params.set("empleado_id", empleadoId)
  if (estado) params.set("estado", estado)
  return apiFetch<SolicitudVacacionesListResponse>(
    `/api/vacaciones?${params}`,
    empresaIdOverride ? { headers: { "X-Empresa-Id": empresaIdOverride } } : {},
  )
}

/** Lista las vacaciones (no canceladas) de un empleado, para su ficha. Endpoint dedicado. */
export async function fetchVacacionesEmpleado(
  empleadoId: string,
): Promise<SolicitudVacacionesListResponse> {
  return apiFetch<SolicitudVacacionesListResponse>(`/api/vacaciones/empleado/${empleadoId}`)
}

export async function fetchVacacion(id: string): Promise<SolicitudVacaciones> {
  return apiFetch<SolicitudVacaciones>(`/api/vacaciones/${id}`)
}

export async function createVacacion(
  data: SolicitudVacacionesCreate,
): Promise<SolicitudVacaciones> {
  return apiFetch<SolicitudVacaciones>("/api/vacaciones", {
    method: "POST",
    body: JSON.stringify(data),
  })
}

export async function cancelarVacacion(id: string): Promise<SolicitudVacaciones> {
  return apiFetch<SolicitudVacaciones>(`/api/vacaciones/${id}/cancelar`, {
    method: "PUT",
  })
}

export async function fetchSaldoVacaciones(
  empleadoId: string,
  empresaIdOverride?: string,
): Promise<SaldoVacaciones> {
  return apiFetch<SaldoVacaciones>(
    `/api/vacaciones/saldo/${empleadoId}`,
    empresaIdOverride ? { headers: { "X-Empresa-Id": empresaIdOverride } } : {},
  )
}

/** Exporta el listado de vacaciones (pdf/excel/csv/word) con los filtros activos aplicados. */
export function exportarVacaciones(
  formato: FormatoExport,
  empresaIdOverride?: string,
  areaId?: string,
  empleadoId?: string,
  estado?: string,
): Promise<void> {
  const headers = empresaIdOverride ? { "X-Empresa-Id": empresaIdOverride } : undefined
  return descargarArchivo("/api/vacaciones/exportar", formato, "vacaciones", headers, {
    area_id: areaId,
    empleado_id: empleadoId,
    estado,
  })
}
